### DISCUSSION TITLE
T_COLOR_FEATURE not working for shaders?

### DISCUSSION THREAD

**Discussion by FrozenCodeException** [Makuna/NeoPixelBus#526] [Q&A] [answered]
Hi community,

I am using SK6812 (RGBW) leds and an arduino mega 2560 R3 for my project. I am looking for a way to use a "global brightness" that scales all pixels in brightness. First i tried NeoPixelBrightnessBus, but then i got over this comment from Makuna (https://github.com/Makuna/NeoPixelBus/issues/339#issuecomment-591066951) where he recommend using a shader. I implemented one similar to the NeoPixelBufferShader-example. 

Before my strip worked just fine with  the "NeoGrbwFeature". After applying the shader (also with NeoGrbwFeature) the colors red and green are switched. Even if i try to use "NeoRgbwFeature" nothing changes. Somebody knows what went wrong here? Is it allowed to use RGBW-LEDstrips with this feature?

greetings 
FrozenCode

**Comment by Makuna** [comment, marked as answer]
It should work.  
Do you have code we can see?  Primarily the shader implementation, the strip and buffer definitions, and where you call to render it.

**Reply by FrozenCodeException** [reply]
Hey Makuna, thanks for the really fast reply. I really appreciate your work in this project!

Here are the important parts from the code, except the template of the shader. This was 1:1 copied from the NeoPixelBufferShader-example (https://github.com/Makuna/NeoPixelBus/blob/a561a8294eb1ba1eaeeafc91467b9ab828dfcfa8/examples/bitmaps/NeoPixelBufferShader/NeoPixelBufferShader.ino).

```
#include <NeoPixelBus.h>
#define NUM_LEDS     445
#define LED_DATA_PIN  11
// template here
NeoPixelBus<NeoGrbwFeature, Neo800KbpsMethod> strip(NUM_LEDS, LED_DATA_PIN);  // could also use NeoSk6812Method? (its a bit faster)
NeoBuffer<NeoBufferMethod<NeoGrbwFeature>> image(NUM_LEDS,1,NULL);
BrightnessShader<NeoGrbwFeature> shader;

void setup(){
  strip.Begin();  
  strip.Show();                                                      
  image.ClearTo(RgbwColor(0,0,0,0));
}

void loop(){
  uint8_t x = 0;
  uint8_t y = 0;
  // strip.SetPixelColor(x++, RgbwColor(255, 0, 0, 0)); // R is Red
  // strip.SetPixelColor(x++, RgbwColor(0, 255, 0, 0)); // G is Green
  // strip.SetPixelColor(x++, RgbwColor(0, 0, 255, 0)); // B is Blue
  image.SetPixelColor(x++, y, RgbwColor(255, 0, 0, 0)); // R is Green
  image.SetPixelColor(x++, y, RgbwColor(0, 255, 0, 0)); // G is Red
  image.SetPixelColor(x++, y, RgbwColor(0, 0, 255, 0)); // B is Blue
  image.Render<BrightnessShader<NeoGrbwFeature>>(strip, shader);
  strip.Show();
  delay(500);
}
```

The comments behind the lines state the colors that are shown if i execute it. So if i use the "imaage" as buffer red and green are swapped. If i comment the four "image"-lines (image and render) and uncomment the "strip"-lines it works fine (of cause without dimming). 
Replacing the "NeoGrbwFeature" everywhere with "NeoRgbwFeature" does not change anything. 

Thanks for your effort in advance.

**Reply by Makuna** [reply]
Quick glance, this looks fine, but I need to see how you implemented the shader.

**Reply by FrozenCodeException** [reply]
The implementation i used was the same as the example. This means: 

```
template<typename T_COLOR_FEATURE> class BrightnessShader : public NeoShaderBase
{
public:
  BrightnessShader():
    NeoShaderBase(),
    _brightness(255) // default to full bright
  {}

  // required for a shader object, it will be called for
  // every pixel
  void Apply(uint16_t index, uint8_t* pDest, uint8_t* pSrc)
  {
    // we don't care what the index is so we ignore it
    //
    // to apply our brightness shader, 
    // use the source color, modify, and apply to the destination 
    
    // for every byte in the pixel,
    // scale the source value by the brightness and 
    // store it in the destination byte
    const uint8_t* pSrcEnd = pSrc + T_COLOR_FEATURE::PixelSize;
    while (pSrc != pSrcEnd)
    {
        *pDest++ = (*pSrc++ * (uint16_t(_brightness) + 1)) >> 8;
    }
  }

  // provide an accessor to set brightness
  void setBrightness(uint8_t brightness)
  {
    _brightness = brightness;
    Dirty(); // must call dirty when a property changes
  }

  // provide an accessor to get brightness
  uint8_t getBrightness()
  {
    return _brightness;
  }
  
private:
  uint8_t _brightness;    
};
```

**Reply by Makuna** [reply]
You found a bug!

I fixed it is in this branch, pick it up and give it a try.  
https://github.com/Makuna/NeoPixelBus/tree/NeoBufferRenderFix

Note:  Apply now requires the source argument to be const, like
```
void Apply(uint16_t index, uint8_t* pDest, const uint8_t* pSrc)
```

**Reply by FrozenCodeException** [reply]
Perfect, works excellent now! 
Thank you very much, have a great day :)
